Puara/RateOfChange.hpp

```diff
@@ -0,0 +1,143 @@
+#pragma once
+#include "halp_utils.hpp"
+
+#include <halp/controls.hpp>
+#include <halp/meta.hpp>
+
+#include <cmath>
+
+#include <cstddef>
+#include <vector>
+
+namespace puara_gestures::objects
+{
+// rate of change with sample count OR time window
+class RateOfChange
+{
+public:
+  //==== window mode ===//
+  enum class WindowMode
+  {
+    SampleCount, ///< Use fixed number of samples
+    TimeWindow   ///< Use fixed time window
+  };
+
+  //==== output units (used by combobox) ===//
+  enum
+  {
+    PerMillisecond,
+    PerSecond,
+    PerMinute,
+    PerHour
+  };
+
+  halp_meta(name, "Rate of change")
+  halp_meta(category, "Analysis/Data")
+  halp_meta(c_name, "rate_of_change")
+  halp_meta(author, "Luana Belinsky")
+  halp_meta(
+      description,
+      "Calculates the rate of change (delta) of an input signal. "
+      "Supports either a fixed sample window OR a time window, and output in different "
+      "time units. "
+      "When using time window mode, if no new data arrives, the last value is held "
+      "constant.")
+  halp_meta(uuid, "7137a3bc-b3c0-43bc-9774-61c0e272db71")
+
+  struct
+  {
+    // Signal input
+    halp::data_port<"Signal", "Int or float. Input signal to analyze", float> signal;
+
+    // Window mode selection using enum_t
+    halp::enum_t<WindowMode, "Window mode"> window_mode{WindowMode::SampleCount};
+
+    // Sample window (N) - visible when WindowMode::SampleCount is selected
+    halp::spinbox_i32<"Sample count", halp::range{2, 1000000, 50}> sample_count;
+
+    // Time window (seconds) - visible when WindowMode::TimeWindow is selected
+    halp::spinbox_f32<"Time window (s)", halp::range{0.001, 3600.0, 1.0}> time_window;
+
+    // Output units using combobox style
+    struct
+    {
+      halp__enum_combobox(
+          "Output units",
+          PerSecond, // default
+          PerMillisecond, PerSecond, PerMinute, PerHour)
+    } output_units;
+
+  } inputs;
+
+  struct
+  {
+    halp::data_port<
+        "Rate of change (delta/time)", "Float. Units as per 'Output units'.", float>
+        out;
+
+    // Diagnostic outputs
+    halp::data_port<
+        "Actual samples in window", "Int. Current number of samples in analysis window.",
+        int>
+        actual_samples;
+    halp::data_port<
+        "Actual window time span (s)",
+        "Float. Actual time span covered by window in seconds.", float>
+        actual_time;
+  } outputs;
+
+  halp::setup setup;
+  void prepare(halp::setup info);
+
+  using tick = halp::tick;
+  void operator()(halp::tick t);
+
+  // derivative across the current window: (x_last - x_first) / sum(dt)
+  float delta_per(int unit) const;
+
+private:
+  // each sample pair contains value, dt, and absolute time
+  struct Sample
+  {
+    float x;   // value
+    double dt; // delta time since previous sample (seconds)
+    double t;  // absolute time since start (seconds)
+  };
+
+  // ring buffer storage
+  std::vector<Sample> _buf;
```

Comment:
I think there's already an unholy amount of different ring buffer implementations in ossia :p 
What do you think about reusing the boost one? https://www.boost.org/doc/libs/latest/doc/html/circular_buffer.html